Some prerequisites for further work

Xentraxx opened the pull request:
Heyho!

Will work on the TODOs now but needed to bump some versions to be able to use modern pub packages.

Just did some necessary breaking changes to the code for the versions I bumped and created a new version with changelog and another TODO.

Your repo seems to be more active than the original one, so I'll merge my changes here.

Cheers

Wacheee commented:
Hey, thanks for this PR! Regarding writing EXIF data to files, I noticed that the code always overwrites the GPS data, even if the media already has this information (many pictures already have this info). Maybe it’d be better to check whether a file already has GPS info before overwriting it. I haven’t tested how much time this adds to the overall process yet, but I’ll do that soon!

Xentraxx commented:
Hey man,

Okay, I know I changed a LOT!
Basically the overview is:
I restructured the project with Steps. Therefore it is easier to understand the flow of what is happening.
And you are absolutely right with " overwrites the GPS data, even if the media already has this information", so I added a check for that. Step 5 (what I implemented) will only run when the "write-exif" option is set.

Otherwise I basically just improved the user experience through more progress bars, replaced some print() with log() (because no one reads through that wall of text anyway), applied lint rules, upgraded dependencies, fixed old tests (which didn't work on windoza) and added new unit tests for the exif stuff.
Also added some TODOs to find stuff which is worth looking into again.

In the current version, all unit tests pass and I ran my (around 100GB) takeout through it (haven't tried the write-exif flag extensively yet).

Now what is left:

- My newly added functionality is painfully slow. I use the decodeNamedImage() and encodeNamedImage() functions from the Image library. I experimented a bit with parallelization but that didn't help much. Maybe the specific en/decoders would be faster but I think it's not that important for now, given that they are behind a clearly as experimental marked flag.

- the decodeNamedImage() and encodeNamedImage() functions from the Image library work fine with jpgs. But png should work as well. It just doesn't. Need to look into that again.

-  For some reason, while moving (without write-exif flag), the application exists. Need to look into this and figure out why that is:
![error](https://github.com/user-attachments/assets/5131863f-9e91-46aa-aff5-fc185b454efc)

- Testing. I need to test everything and make sure it works as intended before you should accept the PR. Probably gonna write more unit tests.

But I am done regarding major changes to the project. I want to get a stable version now which can be merged before I start adding more functionality. The current code needs to be bulletproof first.

jaimetur commented:
Hi @Wacheee, if you are planning to apply massive and not well tested changes like this which modifies the current behavior of the script, please let me know because I need to trust in a reliable and stable version with minor changes in order to minimize the adaptation of CloudPhotoMigrator tool which integrates your forked script since last month. If there is massive changes like this I'm afraid I should go back to the original script because I need more stability with my integration.

Xentraxx commented:
@jaimetur 
No worries. I'll let you guys know once I have a well rounded version and when I am comfortable that everything is working as expected.
But due to the growing complexity of the application I really need to write more tests to catch all those sneaky bugs/edge cases.
Then of course I appreciate some testing by others as well because I can't be 100% sure not to have missed anything.

jaimetur commented:
@Xentraxx all cosmetic changes are welcome for me but I have some concerns about those changes affecting the current behavior of the script such as change some prints by logs as you stated before. Currently I am capturing stdout and stderr outputs from my Tool. If you do some changes here probably the behavior will change when I integrate it within CloudPhotoMigrator tool. 

Also if you do some changes in the interface or parameters it will affect the integration. 

Last month I decided to migrate from the official repository to this fork by @Wacheee because he fixed some recent changes in Google Takeout but ok the other hand he modified the behavior of the year/month output structure option so si had to adapt it to the new arguments.

Wacheee commented:
@Xentraxx
Thank you for the report. I’ve reviewed some of the changes and, apart from what you mentioned, I can see additional updates such as added variable types (which is great), renamed variables/functions, modified some function implementations, and moved several files.

There are quite a few changes, so it would be better to submit multiple smaller pull requests instead of one large one, since that makes testing and review much easier and more controlled.

What we could do is merge these changes into a separate branch (not master) and then create a beta release for people to test. I can test on Windows, but unfortunately I’m unable to test on Linux or macOS right now.

Additionally, the changelogs for this fork are in the `release/exe` branch. You can view the latest changelog here:
https://github.com/Wacheee/GooglePhotosTakeoutHelper/blob/release/exe/CHANGELOG.md?plain=1
This branch is separate from master because initially I wanted to avoid touching the changelogs and other metadata files in `master` to stay in sync with the original fork,

@jaimetur:
The last stable release of this fork is v3.6.2-wacheee: https://github.com/Wacheee/GooglePhotosTakeoutHelper/releases/tag/v3.6.2-wacheee
This pull request will be published as a pre-release (beta) so others can help with testing and give feedback. You can rest assured that a stable release will follow after testing.

Xentraxx commented:
btw @jaimetur your CloudPhotoMigrator looks great! I will check your project in detail once I am done here. It may make sense to implement the upload for nextcloud in your project. It would just be more versatile.

Xentraxx commented:
Alright, everything works as expected so far.
Everything relevant is in the changelog.

@jaimetur if you just read stdout I wouldn't worry. But of course the output has changed if you check for certain output. But it's better now. I used the existing progress bars and the steps provide even more standardisation. 

Sometimes at around 80% of Step 7, the program exits with "Exited (-1073740940)." which is a Heap corruption error.
That can't really be caused by my changes in business logic but only the dart version.
We should definitely try some larger exports and if any of you also get this error, we might have to play around with the dart version/dependencies again. Otherwise I am afraid I have no idea how to fix this, as it is a bit too low level...
But it might also just be a local issue on my side.

I will look into creating a new branch and smaller pull requests tomorrow.

But I am afraid due to all the formatting and lint refactoring changes, you'd have to go through everything anyway.

This is the final 4.0.0 release candidate.

Cheers!

Xentraxx commented:
As discussed, here is the new pull request in a new beta branch: https://github.com/Wacheee/GooglePhotosTakeoutHelper/pull/13
